File: src/main/java/net/nutchi/anyban/command/BanIpCommand.java
package net.nutchi.anyban.command;

import net.md_5.bungee.api.CommandSender;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.plugin.Command;
import net.md_5.bungee.api.plugin.TabExecutor;
import net.nutchi.anyban.AnyBan;
import net.nutchi.anyban.util.Message;
import net.nutchi.anyban.util.IpChecker;

import java.net.InetSocketAddress;
import java.util.Arrays;
import java.util.Collections;

public class BanIpCommand extends Command implements TabExecutor {
    private final AnyBan plugin;

    public BanIpCommand(AnyBan plugin) {
        super("aban-ip", "anyban.command.ban-ip");
        this.plugin = plugin;
    }

    @Override
    public void execute(CommandSender sender, String[] args) {
        if (args.length >= 1) {
            String target = args[0];
            String reason = args.length == 1 ? Message.DEFAULT_BAN_REASON.text() : String.join(" ", Arrays.copyOfRange(args, 1, args.length));
            String source = sender instanceof ProxiedPlayer ? sender.getName() : Message.CONSOLE_SOURCE.text();

            ProxiedPlayer proxiedPlayer = plugin.getProxy().getPlayer(target);

            String ip;
            if (IpChecker.isIp(target)) {
                ip = target;
            } else if (proxiedPlayer != null) {
                ip = ((InetSocketAddress) proxiedPlayer.getSocketAddress()).getAddress().getHostAddress();
            } else {
                ip = null;
            }

            if (ip != null) {
                if (!plugin.getBannedIpManager().isBanned(ip)) {
                    plugin.getBannedIpManager().add(ip, source, reason);

                    sender.sendMessage(Message.BAN_IP_COMMAND.component(ip, reason));

                    plugin.getProxy().getPlayers().forEach(p -> {
                        String pIp = ((InetSocketAddress) p.getSocketAddress()).getAddress().getHostAddress();
                        if (pIp.equals(ip)) {
                            p.disconnect(Message.YOU_ARE_BANNED_IP.component(reason));
                        }
                    });
                } else {
                    sender.sendMessage(Message.IP_ALREADY_BANNED.component());
                }
            } else {
                sender.sendMessage(Message.INVALID_IP.component());
            }
        } else {
            sender.sendMessage(Message.INCOMPLETE_COMMAND.component());
        }
    }

    @Override
    public Iterable<String> onTabComplete(CommandSender sender, String[] args) {
        return Collections.emptyList();
    }
}

File: src/main/kotlin/com/github/tsuoihito/anyban/commands/BanIpCommand.kt
package com.github.tsuoihito.anyban.commands

import com.github.tsuoihito.anyban.AnyBan
import com.github.tsuoihito.anyban.handlers.getAddedBannedIpList
import com.github.tsuoihito.anyban.handlers.getPlayersCandidate
import com.github.tsuoihito.anyban.handlers.getUserIp
import com.github.tsuoihito.anyban.utils.calcCurrentTimeString
import com.github.tsuoihito.anyban.utils.getMessage
import com.github.tsuoihito.anyban.utils.kickPlayer
import net.md_5.bungee.api.CommandSender
import net.md_5.bungee.api.plugin.Command
import net.md_5.bungee.api.plugin.TabExecutor
import java.util.concurrent.CompletableFuture

class BanIpCommand(private val plugin: AnyBan) : Command("aban-ip", "anyban.command.ban-ip"), TabExecutor {
    override fun execute(sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            sender.sendMessage(getMessage(plugin.config, "message.invalidSyntax"))
            return
        }
        val name = args[0]
        val reason = if (args.size > 1) args[1] else plugin.config?.getString("defaultBanReason").orEmpty()

        val ips = getUserIp(plugin.userIpList, name)?.ips
        if (ips == null) {
            sender.sendMessage(getMessage(plugin.config, "message.ipNotFound") { it.replace("%name%", args[0]) })
            return
        }
        ips.forEach {
            plugin.bannedIpList =
                getAddedBannedIpList(plugin.bannedIpList, it, calcCurrentTimeString(), sender.name, "forever", reason)
        }

        sender.sendMessage(getMessage(plugin.config, "message.banSuccess") { it.replace("%name%", name) })
        kickPlayer(
            plugin.proxy,
            name,
            getMessage(plugin.config, "message.youAreBanned", false) { it.replace("%reason%", reason) })

        CompletableFuture.runAsync { plugin.saveData() }
    }

    override fun onTabComplete(sender: CommandSender, args: Array<String>): Iterable<String> {
        return if (args.size == 1) {
            getPlayersCandidate(plugin.userUuidList, args[0])
        } else listOf()
    }
}
